csharp/src/Ice/InputStream-Class.cs

```diff
@@ -311,90 +359,217 @@ private AnyClass ReadInstance(int index)
-                // not created yet.
+                // Add the instance to the map/list of instances. This must be done before reading the instances (for
+                // circular references).
+                _instanceMap.Add(instance);
 
-                IClassFactory? factory = null;
-                if (typeIdOpt is string typeId)
+                // Read all the deferred indirection tables now that the instance is inserted in _instanceMap.
+                if (_current.DeferredIndirectionTableList11?.Count > 0)
                 {
-                    Debug.Assert(_current.SliceCompactId == null);
-                    factory = Communicator.FindClassFactory(typeId);
+                    int savedPos = _pos;
+
+                    Debug.Assert(_current.Slices?.Count == _current.DeferredIndirectionTableList11.Count);
+                    for (int i = 0; i < _current.DeferredIndirectionTableList11.Count; ++i)
+                    {
+                        int pos = _current.DeferredIndirectionTableList11[i];
+                        if (pos > 0)
+                        {
+                            _pos = pos;
+                            _current.Slices[i].Instances = Array.AsReadOnly(ReadIndirectionTable());
+                        }
+                        // else remains empty
+                    }
+
+                    _pos = savedPos;
                 }
-                else if (_current.SliceCompactId is int compactId)
+
+                if (readIndirectionTable)
                 {
-                    factory = Communicator.FindClassFactory(compactId);
+                    ReadIndirectionTableIntoCurrent();
                 }
+            }
+            else
+            {
+                // With the 2.0 encoding, we don't need a DeferredIndirectionTableList because all the type IDs are
+                // provided by the first slice (when using the sliced format).
 
-                if (factory != null)
+                (string[]? allTypeIds, _) = ReadFirstSliceHeaderIntoCurrent20();
+                if (allTypeIds != null)
                 {
-                    instance = factory.Read(this);
+                    int skipCount = 0;
+                    foreach (string typeId in allTypeIds)
+                    {
+                        if (Communicator.FindClassFactory(typeId) is IClassFactory factory)
+                        {
+                            instance = factory.Create();
+                            break; // foreach
+                        }
+                        else
+                        {
+                            skipCount++;
+                        }
+                    }
+
+                    instance ??= new UnknownSlicedClass();
+
+                    _instanceMap.Add(instance);
+                    ReadIndirectionTableIntoCurrent(); // read the indirection table immediately
```

Comment:
Is it correct to read the indirection table from the first slice of instance before reading the indirection tables of the skipped Slices? Won't this mess up the instanceMap indexes?

I was expecting to see this call after the for loop... and eventually actually move it to `InputStream.IceReadStartFirstSlice` as suggested by my previous comment?

Reply:
We always read the indirection tables in order (just as they are written), from first (most derived) slice to least derived slice. So yes, it's correct. I think the confusion here is with nested indirection tables - when a class instance in an indirection table has itself class data members with indirection tables: in this case, we read the deepest indirection tables first.

With the 2.0 encoding, the indirection table read immediately after `_instanceMap.Add` is the indirection table for the first (most derived) slice.
I'd rather keep all this indirection table reading processing for the first slice in a single method, `ReadInstance`.